010.md

```diff
@@ -0,0 +1,466 @@
+```javascript
+y => { root: A, branch: [X] }
+```
+
+This message is not well connected, but it's common to render it in later steps
+regardless
+
+
+## 4. Rendering a thread
+
+Each message is generally displayed showing:
+- author name + image (derived from `msg.value.author` and other messages outside
+  the scope of this spec)
+- `timestamp`
+- `text` rendered as markdown
+- sometimes further data such as
+  - a summary of `vote` responses
+  - info about any `fork` that this `post` started
+  - a list of other threads which have linked to this message (backlinks)
+
+### 4.1. Linear rendering of a tangle
+
+A tangle with forks is hard to render as a nice linear output (the most familiar
+way humans like to engage with threads in).
+
+We build a tangle, to get partial ordering, then for **concurrent messages**, we
+break ties using the "timestamp" that the message was published at.
+(timestamp here is the minimum of the asserted timestamp (`msg.value.timestamp)`
+and the timestamp the message was received at (`msg.timestamp` in the JS
+implementation)
+
+
+```mermaid
+flowchart RL
+A(A):::root;B(B);C(C);D(D);M(M);N(N);Y(Y):::disconnected
+N-->M-->C & D-->B-->A
+Y
+
+
+D-.->C
+N-.->Y-.->M
+linkStyle 6,7,8 stroke:#53e9d4,stroke-width:4;
+
+classDef default fill: #e953da, stroke-width: 0, color: white
+classDef root fill: #4c44cf, stroke-width: 0, color: white
+classDef disconnected stroke:none,fill:#e95353,color:none;
+```
+_Figure. showing the creation of a tangle, then tie-breaking between concurrent
+messages `C`,`D`, and best-guess insertion of disconnected message `Y`_
+
+For **disconnected messages** we try to place them in the correct position by
+timestamp alone. It could be a good idea to indicate in UI that this message may
+be in the wrong place and is missing context.
+
+
+```mermaid
+flowchart RL
+A(A):::root;B(B);C(C);D(D);M(M);N(N);Y(Y):::disconnected
+N-->Y-->M-->D-->C-->B-->A
+classDef default fill: #e953da, stroke-width: 0, color: white
+classDef root fill: #4c44cf, stroke-width: 0, color: white
+classDef disconnected stroke:none,fill:#e95353,color:none;
+```
+_Figure. showing a linearised thread tangle, along with a "best guess" placement
+of disconnected message `Y`._
+
+
+### 4.2. Scuttlebutt Markdown
+
+The `text` field of `post` messages is Scuttlebutt flavoured Markdown, which is
+basic Markdown with a couple of important privacy enhancing features.
+
+#### 4.2.1 Image rendering
+
+DO NOT render images that link to the web. This leaks data, which allows
+tracking of users.
+```markdown
+![unsafe image](https://upload.wikimedia.org/wikipedia/commons/1/11/Panopticon.jpg)
+```
+Such images MUST instead be rendered as a link with an alert/ warning
+(see 4.2.2).
+
+ONLY render images that link to blobs that are hosted locally:
+```markdown
+![snails](&7AUsuVcxUzIJZPBBLqR3O2KvIiuECgdYq9lzEh0jcS8=.sha256)
+```
+
+#### 4.2.2 Link rendering
+
+When a link to the web is present, the interface MUST render an alert to users
```

Comment:
```suggestion
When a link to the web is present, the interface SHOULD render an alert to users
```

I don't actually think an alert is important. In Manyverse we use black underscore links for WWW links and blue underscore links for SSB links and that's already one way of signaling to users.

Reply:
It was too subtle for me to know! Good to know though. Maybe www should be RED